espnet3/runner/base_runner.py

```diff
@@ -0,0 +1,392 @@
+
+        Keep this as a ``@staticmethod`` so that it is pickle-safe for Dask
+        and does not capture ``self``.
+
+        Args:
+            idx (int): The input index to process.
+            dataset: Dataset object provided via the environment.
+            model: Model object provided via the environment.
+            **env: Any additional environment entries injected by the provider.
+
+        Returns:
+            Any: Result for the given index.
+
+        Raises:
+            NotImplementedError: Always in the base class; implement in subclass.
+
+        Example:
+            >>> class MyRunner(BaseRunner):
+            ...     @staticmethod
+            ...     def forward(idx, *, dataset, model, **env):
+            ...         x = dataset[idx]
+            ...         return model(x)
+        """
+        raise NotImplementedError
+
+    def _run_local(self, indices: Sequence[int]) -> List[Any]:
+        """Run sequentially on the driver using a locally built environment.
+
+        Args:
+            indices (Sequence[int]): Indices to process.
+
+        Returns:
+            List[Any]: Results in the same order as ``indices``.
+
+        Notes:
+            - Uses ``tqdm`` progress bar over the input sequence.
+        """
+        env = self.provider.build_env_local()
+        f = self.__class__.forward  # static
+        return [f(i, **env) for i in tqdm(indices, total=len(indices))]
+
+    def _run_parallel(self, indices: Sequence[int]) -> List[Any]:
+        """Run with synchronous Dask mapping using per-worker environments.
+
+        Args:
+            indices (Sequence[int]): Indices to process.
+
+        Returns:
+            List[Any]: Results gathered in order.
+
+        Notes:
+            - Wraps ``forward`` with :func:`wrap_func_with_worker_env` so that
+              missing keyword args are injected from the worker env.
+        """
+        setup_fn = self.provider.make_worker_setup_fn()
+        out = []
+        with get_client(get_parallel_config()) as client:
+            for res in parallel_for(
+                self.__class__.forward, indices, setup_fn=setup_fn, client=client
+            ):
+                out.append(res)
+        return out
+
+    async def _run_async(self, indices: Sequence[int]) -> List[Any] | None:
+        """Submit shards to a Dask cluster and gather or write results.
+
+        Workflow:
+            1) Emit per-shard JSON specs to ``async_specs_dir``.
+            2) Submit worker entrypoints that reconstruct runner/provider and run.
+            3) Write results to ``async_result_dir``.
+        Args:
+            indices (Sequence[int]): Indices to process.
+
+        Returns:
+            List[Any] | None: Flattened results if gathered; otherwise ``None``.
+
+        Raises:
+            RuntimeError: If Dask client creation fails.
+            ValueError: If async sharding parameters are invalid.
```

Comment:
Will Timeout or other errors (e.g., caused when long-time jobs are automatically killed or killed due to OOM) also be handled?

Reply:
Oh, thank you, for async mode it is simpler and easier I believe, as all workers are independent job and all logs are in their log files...

Do you think if we need detailed loggings? I think we can put detailed logs to help debugging, but I think we can create them in another PR to put loggings in all packages